app/app.go

```diff
@@ -68,11 +68,10 @@ func (app *App) RunPlan(ctx context.Context, stateStorage storage.Storage, plan
 
 	outStreamID := &execution.StreamID{Id: "output"}
 
-	pullEngine := execution.NewPullEngine(out, stateStorage, stream[0], outStreamID, execOutput[0].WatermarkSource, true)
+	ctx, cancel := context.WithCancel(ctx)
+	pullEngine := execution.NewPullEngine(out, stateStorage, stream[0], outStreamID, execOutput[0].WatermarkSource, true, cancel)
```

Comment:
Give the pull engine a context, then do context.WithCancel(ctx) in there. And then check ctx.Done on it. (you can remove the ctx from pullEngine.Run())